GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/LooseObjectStepTests.cs

```diff
@@ -0,0 +1,139 @@
+﻿using GVFS.FunctionalTests.FileSystemRunners;
+using GVFS.FunctionalTests.Tools;
+using NUnit.Framework;
+using System.Collections.Generic;
+using System.IO;
+using System.Text.RegularExpressions;
+
+namespace GVFS.FunctionalTests.Tests.EnlistmentPerFixture
+{
+    // TODO(Mac): Before these tests can be enabled PostFetchJobShouldComplete needs
+    // to work on Mac (where post-fetch.lock is not removed from disk)
+    [TestFixture]
+    [Category(Categories.FullSuiteOnly)]
+    [Category(Categories.MacTODO.M4)]
+    public class LooseObjectStepTests : TestsWithEnlistmentPerFixture
+    {
+        private const string TempPackFolder = "tempPacks";
+        private FileSystemRunner fileSystem;
+
+        // Set forcePerRepoObjectCache to true to avoid any of the tests inadvertently corrupting
+        // the cache 
+        public LooseObjectStepTests()
+            : base(forcePerRepoObjectCache: true)
+        {
+            this.fileSystem = new SystemIORunner();
+        }
+
+        private string GitObjectRoot
+        {
+            get
+            {
+                return this.Enlistment.GetObjectRoot(this.fileSystem);
+            }
+        }
+
+        private string PackRoot
+        {
+            get
+            {
+                return this.Enlistment.GetPackRoot(this.fileSystem);
+            }
+        }
+
+        private string TempPackRoot
+        {
+            get
+            {
+                return Path.Combine(this.PackRoot, TempPackFolder);
+            }
+        }
+
+        [TestCase, Order(1)]
+        public void RemoveLooseObjectsInPackFiles()
+        {
+            // Delete any starting loose objects and verify
+            this.DeleteFiles(this.GetLooseObjectFiles());
+            Assert.AreEqual(0, this.GetLooseObjectFiles().Count);
+
+            // Move packfiles to temp
+            this.MovePackFilesToTemp();
+
+            // Expand 1 pack file and Copy it back to packs
+            this.ExpandOneTempPackAndMoveBack();
+
+            // Verify we have some LooseObjects
+            // These objects will also appear in the pack file that was moved back
+            Assert.AreNotEqual(0, this.GetLooseObjectFiles().Count);
+
+            // Run Cleanup
+            this.Enlistment.LooseObjectStep();
+
+            // Verify loose objects appearing in the pack file are removed
+            Assert.AreEqual(0, this.GetLooseObjectFiles().Count);

```

Comment:
Can we run an add or commit and create some objects that would be left so we can verify that some objects don't get removed or would that be a separate test?

Reply:
Seperate test.  And there are definitely more functional tests coming after this PR :-).  Trying to keep the scope narrow here as the PR is already huge.